pkg/proxy/ipvs/proxier.go

```diff
@@ -1591,7 +1576,7 @@ func (proxier *Proxier) createAndLinkeKubeChain() {
 			if chain, ok := existingNATChains[ch.chain]; ok {
 				writeBytesLine(proxier.natChains, chain)
 			} else {
-				writeLine(proxier.natChains, utiliptables.MakeChainLine(kubePostroutingChain))
+				writeLine(proxier.natChains, utiliptables.MakeChainLine(ch.chain))
```

Comment:
I think this code is copy form [iptables/proxier.go#L722-L730](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L722-L730), it means if chain not exist, create it.

Reply:
I'm not sure what the original intent here was (maybe @m1093782566 knows), but this does seem wrong but I don't think it broke anyone because we already ensure the chain exists in line 1605.

Reply:
I write some test code, [iptables_test1.go](https://gist.github.com/liuxu623/cd0056e6baec485cfa91811c0618b2b4) and [iptables_test2.go](https://gist.github.com/liuxu623/d953c27e33dc8b71938587c2bc21e8c8), when all kube iptables chains not exist, iptables_test1.go output is 
```
:KUBE-POSTROUTING - [0:0]
:KUBE-POSTROUTING - [0:0]
:KUBE-POSTROUTING - [0:0]
:KUBE-POSTROUTING - [0:0]
```
iptables_test2.go output is
```
:KUBE-SERVICES - [0:0]
:KUBE-FIREWALL - [0:0]
:KUBE-NODE-PORT - [0:0]
:KUBE-LOAD-BALANCER - [0:0]
```

Reply:
You are right, we already ensure the chain exists in line 1605, so I think we don't actually need line1609-1622 anymore.

Reply:
We can do this in iptables proxier too, use iptables.EnsureChain to create not exist chains instead of iptables.RestoreAll.

Reply:
Letting `RestoreAll` ensure the chains is more efficient than making extra `EnsureChain` calls. We should only call `EnsureChain` for the chains that we can't ensure from `RestoreAll` because we aren't filling in their *contents*. (eg, the `RestoreAll` input might include `-j KUBE-MARK-DROP` in some rule, but we can't actually define the `KUBE-MARK-DROP` chain in `RestoreAll` because we don't know exactly what the rule in that chain is supposed to be. So we should `EnsureChain` it.)

Reply:
ok, so we should remove `EnsureChain` ? Becase we already use `RestoreAll ` to ensure the chains exists.